### ISSUE TITLE
A nullable `BuildContext` can trick `use_build_context_synchronously`

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4342]
@stuartmorgan found this issue. It boils down to a simple case like this:

```dart
import 'package:flutter/widgets.dart';

void foo(BuildContext? context) async {
  await f();
  Navigator.of(context!);
}

Future<void> f() async {}
```

`use_build_context_synchronously` should be reported in this code, but `context` isn't _exactly_ a `BuildContext`; it's a `BuildContext?` so it slips past the rule. This was found in code in the image_picker Flutter plugin.
